### ISSUE TITLE
Init method is now 6 time slower than an old one (from July 2019) (6sec vs 1.2 sec)

### ISSUE THREAD

**Issue by Moumouls** [herumi/she-wasm#9]
When the lib switched from wasm file to data uri system, a js blocking loop issue was introduced.

![Capture d’écran 2021-05-12 à 15 11 07](https://user-images.githubusercontent.com/27959372/117980904-b9189780-b334-11eb-8e1d-cb233e7d77ab.png)

The issue is reproducilble on `/browser/she-demo.html`

I am trying to find the source and a possible solution. 

@herumi may be you have an idea of improvement to keep the nice and simple data uri system without blocking the JS loop (async data uri loading).

Note: i'll post my progress here, if i found something.

**Comment by Moumouls** [comment]
Found that the `Webassembly.instantiate` has some performance drawbacks.
https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/instantiate

Also may be the issue is out of she wasm scope since `she_c.js` is auto generated by `emcc`.

Here an emcc improvement could be trying to stream the data uri and simulate a response.

```js
 function instantiateArrayBuffer(receiver) {
  return getBinaryPromise().then(function(binary) {
  // Something like this
   return WebAssembly.instantiateStreaming(new Response(binary,init), info);
  }).then(receiver, function(reason) {
   err("failed to asynchronously prepare wasm: " + reason);
   abort(reason);
  });
 }
```

**Comment by Moumouls** [comment]
attempted to build she-wasm with latest emcc (2.0.20-git) and then build browser she.js, the issue persist.

**Comment by Moumouls** [comment]
Okay so data uri system looks like ok, but something went wrong in `_sheInit` and `_sheSetRangeForDLP` both methods now block the main JS loop and take 3sec each to return

**Comment by Moumouls** [comment]
Ok after a rollback to commit `022e46efc4a3aa97ef049216c215ed2df47d0197` demo is working properly, and `_sheInit` + `_sheSetRangeForDLP` return in 1.2sec.

![Capture d’écran 2021-05-12 à 18 57 48](https://user-images.githubusercontent.com/27959372/118015076-2ee02b80-b354-11eb-9ad2-a8ac2c5b0caa.png)

So the new wasm seems 6 time slower at init time than the old one (17 July 2019).

Since i'm not an cpp/emcc expert, i'll wait your feedback @herumi  before some other investigations 🙂

**Comment by herumi** [comment]
Okay, I'll check it later.

**Comment by herumi** [comment]
This issue seems to have been caused by https://github.com/herumi/she-wasm/commit/97e3bd451747bce92f182a74236d1f9c828bc8d0 .
she_c.wasm was combined into she_c.js by this commit.
I don't know why it causes the speed down.
I'll investigate it more.

**Comment by Moumouls** [comment]
Hi @herumi  thanks for your feedback, one thing that i noticed is the performance warning on MDN docs about the `WebAssembly.instantiate()` used by the new base64 system.

But at runtime the `WebAssembly.instantiate()` method looks like to resolve pretty much faster (similar speed when instantiateStreaming was used.)

This afternoon (French hour) i'll try to compile the she wasm with the old 2 file system. May be there is an issue with the `instanciate()` method at runtime on the first function execution.

**Comment by herumi** [comment]
How do you want to use she.js?
If you use she.js on browsers without webpack/react, then I'll make a split version to she_c.js and she_c.wasm.

**Comment by Moumouls** [comment]
Hi @herumi i failed to find some time to try the split version. I'll try this directly this morning.

Currently we use this lib in NodeJS and also in browser under a NextJS app (React/Webpack).

The old split version system worked well, but it was difficult to provide a custom she.wasm location without diving into the auto generated JS script (also a fork was needed).

Is there an option on `emcc` to expose the wasm location, or provide it at init time ?

**Comment by herumi** [comment]
>Is there an option on emcc to expose the wasm location, or provide it at init time ?

I could not find it when I looked for it before. I'll do it again.

According to https://github.com/herumi/bls-eth-wasm/issues/4 , a single file is easy to use.

**Comment by Moumouls** [comment]
Okay here result on cross demo with split version, there a 3-4sec blocking UI effect, (3 time slower than the old version)

![Capture d’écran 2021-05-21 à 10 26 53](https://user-images.githubusercontent.com/27959372/119107033-2a380900-ba1f-11eb-9644-d1562dfd911d.png)

Here a PR with the tested perf: https://github.com/herumi/she-wasm/pull/10

**Comment by Moumouls** [comment]
On cross demo with single file: 

![Capture d’écran 2021-05-21 à 10 33 37](https://user-images.githubusercontent.com/27959372/119108042-150faa00-ba20-11eb-903b-a999a1c50cf3.png)

So here the issue seems to not be the single file feature.

**Comment by Moumouls** [comment]
So here, you are right it's better to keep sticked to the single file (way more easy to use system), but something strange happened into the wasm binary at compilation time.

The new she_c.wasm has a weight of 500ko and the old one 456 ko. I know that you have implemented some new methods/feats on she_wasm (dec result and others), but a 44ko diff seems big to me.

May be something was imported into the cpp files and lead to a slowed down init ? (sorry if my remark is not relevant, I don't have much knowledge about the cpp part)

**Comment by Moumouls** [comment]
Okay ater watching the perf sreen shot:

Old init picture:
![Capture d’écran 2021-05-21 à 10 49 50](https://user-images.githubusercontent.com/27959372/119110393-6456da00-ba22-11eb-8527-98f3ae3f3aa1.png)

New init picture:
![Capture d’écran 2021-05-21 à 10 49 44](https://user-images.githubusercontent.com/27959372/119110424-6e78d880-ba22-11eb-8f9a-24f6dee2ef23.png)

it seems that the level 3 of wasm call stack has some block with a doubled execution time

I will get the exact numbers

**Comment by Moumouls** [comment]
the level 3 call stack seems to have a perf issue.

Old:
![Capture d’écran 2021-05-21 à 10 56 28](https://user-images.githubusercontent.com/27959372/119111377-4d64b780-ba23-11eb-8e62-be5b70a9cd27.png)

New:
![Capture d’écran 2021-05-21 à 10 53 32](https://user-images.githubusercontent.com/27959372/119111429-58b7e300-ba23-11eb-8380-813eb04e8cdf.png)

Following the execution, these blocks should be the same.

**Comment by Moumouls** [comment]
So @herumi i just missed that i can export charts. You can load it under chrome dev tool profile tab.

Here the ZIP with Old and new profile if you want to investigate call stack.

**Comment by herumi** [comment]
Thank you for the detailed analysis. I'll check it.

**Comment by Moumouls** [comment]
I hope it will help. This issue is quite annoying, it is quite difficult to bypass it and the blocking of the JS loop does not allow to properly manage the wait on the UI for the end user during the 4-6sec.

I know that a solution exists with web worker but it's not an easy solution.

An additional note here, on our production servers we encounter some blocking JS loop issue during intensive computations. The only solution that we found is to use the dedicated side child process to delegate homo/she_wasm ops.

A similar system could be achieved on browser with web workers.

May be i'll try to find some time in next few months to try to implement a controller working out of the box to support all she ops within a dedicated child process in nodeJS and web worker in browser env. It could be easy since all entities have a serialized version (for cross process communication), but it need more investigation and out of scope here 🙂

**Comment by herumi** [comment]
I found the cause of the slowdown.
The default window size to compute scalar multiplication is 10.
It is okay for PC but it is too big for wasm.
https://github.com/herumi/mcl/blob/master/include/mcl/she.hpp#L39

I've removed the flag to merge js and wasm.
I specified the flag at https://github.com/herumi/she-wasm/commit/70285169fcee26912751e4ac9a358f183c6f99b7 .

**Comment by Moumouls** [comment]
Thanks @herumi i can confirm on my side that init perf is now fast (1sec) (and faster than before).
Do you have an ETA for the new npm version ?

![Capture d’écran 2021-05-25 à 10 37 55](https://user-images.githubusercontent.com/27959372/119467080-806dba80-bd45-11eb-903a-8767d6beb4ee.png)

**Comment by herumi** [comment]
I published v0.9.1 .
https://www.npmjs.com/package/she-wasm
